doc/tcti-spi.md

```diff
@@ -0,0 +1,246 @@
+    t.length = cnt*8;
+    t.tx_buffer = data_out;
+    t.rx_buffer = data_in;
+
+    // Transmit
+    esp_err_t ret = spi_device_polling_transmit(platform_data->spi, &t);
+    if (ret != ESP_OK) {
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    return TSS2_RC_SUCCESS;
+}
+
+// Sleeps for the specified amount of milliseconds
+void platform_sleep_ms (void* user_data, int32_t milliseconds)
+{
+    // Sleep the specified amount of milliseconds
+    vTaskDelay(milliseconds / portTICK_PERIOD_MS);
+}
+
+// Start a timeout timer with specified expiry in milliseconds
+void platform_start_timeout (void* user_data, int32_t milliseconds)
+{
+    // Cast our user data
+    PLATFORM_USERDATA* platform_data = (PLATFORM_USERDATA*) user_data;
+
+    // Store timeout expiry time
+    platform_data->timeout_expiry = esp_timer_get_time() + (milliseconds*1000);
+}
+
+// Check if the timeout timer started by platform_start_timeout has already expired
+bool platform_timeout_expired (void* user_data)
+{
+    // Cast our user data
+    PLATFORM_USERDATA* platform_data = (PLATFORM_USERDATA*) user_data;
+
+    // Check if timeout already expired
+    return esp_timer_get_time() < platform_data->timeout_expiry;
+}
+
+// Frees the platform user_data struct and everything inside
+void platform_finalize(void* user_data)
+{
+    // Cast our user data
+    PLATFORM_USERDATA* platform_data = (PLATFORM_USERDATA*) user_data;
+
+    // Free resources inside user_data like SPI device handles here
+    // ...
+    
+    // Free user_data
+    free(platform_data);
+}
+
+// Creates a new platform struct with the given SPI device handle and chip select
+TSS2_TCTI_SPI_PLATFORM create_tcti_spi_platform(spi_device_handle_t spi, gpio_num_t cs_pin)
+{
+    // Create required platform user data
+    PLATFORM_USERDATA* platform_data = malloc(sizeof(PLATFORM_USERDATA));
+    platform_data->spi = spi;
+    platform_data->cs_pin = cs_pin;
+    platform_data->timeout_expiry = 0;
+    
+    // Create TCTI SPI platform struct with custom platform methods
+    TSS2_TCTI_SPI_PLATFORM platform;
+    platform.user_data = platform_data;
+    platform.sleep_ms = platform_sleep_ms;
+    platform.start_timeout = platform_start_timeout;
+    platform.timeout_expired = platform_timeout_expired;
+    platform.spi_acquire = platform_spi_acquire;
+    platform.spi_release = platform_spi_release;
+    platform.spi_transfer = platform_spi_transfer;
+    platform.finalize = platform_finalize;
+    
+    return platform;
+}
+
+// Creates a SPI TCTI using the example pin defines
+TSS2_TCTI_CONTEXT* create_tcti_ctx()
+{
+    // SPI bus & devive configuration
+    spi_bus_config_t bus_cfg = {
+        .miso_io_num = PIN_NUM_MISO,
+        .mosi_io_num = PIN_NUM_MOSI,
+        .sclk_io_num = PIN_NUM_CLK,
+        .quadwp_io_num = -1,
+        .quadhd_io_num = -1,
+        .max_transfer_sz = 64
+    };
+    spi_device_interface_config_t dev_cfg = {
+        .clock_speed_hz = 10*1000, // 10 KHz
```

Comment:
just an example, but 10khz is quite low.

Reply:
Yes, but a higher SPI clock speed occasionally produced bad SPI transfer results on the demo setup using ~15cm jumper cables between the ESP32 and the TPM (parasitic capacitance?) when 10Khz worked reliably. On a circuit board the speed could be a lot higher.
What is your suggestion for an example clock speed?

Reply:
Should I just add the supported clock speed range as a comment in the example config?

Reply:
Hmm,
so basically we have seen a lot of spi masters with 'broken' frequencies - so it's a bit hard to tell - 10khz sounds really low though. did you try to scope higher frequencies to see whether it's an electrical issue or a controller/master issue?
I just also want to rule out that for whatever reason, some sleeps/timeouts are missing or incorrect which are only triggered at higher frequencies.

Most TPMs run up to 33mhz. On my rpi overlay I specified 32 mhz.
For a microcontroller I think 1-10 Mhz should be a suitable range for spi.

But of course cable length and overspeaking can be an issue.

I'll try to test your commit using the letstrust arduino adapter (https://buyzero.de/products/arduino-adapter-for-letstrust-tpm)  and my esp32 with arduino form factor and see what frequency works on a sensible connection.

Reply:
you can keep it as 10khz, if this is your tested speed, but maybe add a comment about up to 33mhz max

Reply:
Update: I just tried to increase the clock rate up to 22MHz (the ESP32 does not support more using SPI on regular GPIO pins in full-duplex mode) and **it works**!
I just had to make a little fix in `Tss2_Tcti_Spi_Init` because at 22MHz and without logging (which slowed down the ESP just enough that it worked, but without logging it broke) we are too fast for the the TPM after claiming the locality. So now we wait for the STS_COMMAND_READY flag to be set after claiming the locality. And now it works reliably at 22MHz with and without logging.

Reply:
Too fast for the TPM ? Probably not - the particular TPM is testes with 33Mhz+
Maybe your retry-loop is too fast and something expiring.
But yeah you fixed it.

As I said, sometimes the masters also have issues with certain clock rates.
22MHZ sound fine - 10 are probably less stress with the ESP :P

Reply:
Okay, so in the example in `doc/tcti-spi.md` I will specify 10MHz but comment that it has been tested up to 22MHz.